[HTTP 이해 - 웹 서버 구현] 1단계 미션 제출합니다.

bigpel66 opened the pull request:
1단계 미션의 기능 중에서 POST 요청을 이용한 회원가입 (WAS), GET 요청을 이용한 회원가입 (WS)의 컨트롤러 작성만 남겨둔채로 기한 내에 1차 제출을 위해 PR 보내게 되었습니다.

회원 가입 기능을 제외한 기능 요구사항인 
* GET /index.html 응답하기
* CSS 지원하기
* Query String 파싱
* Redirect

기능은 완료된 상태입니다.

<img width="1840" alt="스크린샷 2024-04-09 오후 5 53 26" src="https://github.com/next-step/jwp-was-kakao/assets/47660338/a1274f6b-e688-4499-a57a-df0d7bb0a1bf">

요구사항은 아니었지만, Not Found가 발생한 경우 NOT FOUND 문구를 보게끔 설정해둔 상태입니다.

<img width="1840" alt="스크린샷 2024-04-09 오후 5 53 47" src="https://github.com/next-step/jwp-was-kakao/assets/47660338/f41626f9-f70e-448f-98f3-0b6f03205e3c">

개인적으로 아쉬운 점은
* CustomException을 활용하지 않았던 점
* 컨트롤러 작성까지 미션 기간 내에 달성하지 못했던 점
* Response가 Renderer의 역할을 맡은 점 (원래는 StatusLine, HttpHeaders, ResponseBody를 갖게할 예정이었습니다.)

등이 기억이 납니다.

피드백 해주시는 부분들과 피드백 받기 전까지 구현하면서 회고를 통해 보완할 점 찾아서 잘 반영해보겠습니다.

리뷰 미리 감사드립니다.

htrucci reviewed (changes requested):
리뷰가 늦어 죄송합니다 ㅠ_ㅠ 🙏 
마지막 단계인 세션을 잘 구현해주시고 관련 테스트 코드도 꼼꼼히 짜신 것 같습니다.
몇가지 코멘트 남겨놓았고 /user/list 접근시 model 셋팅이 안되고있어 재확인 부탁드릴게요~!

dongjoon1251 commented:
> 리뷰가 늦어 죄송합니다 ㅠ_ㅠ 🙏
> 마지막 단계인 세션을 잘 구현해주시고 관련 테스트 코드도 꼼꼼히 짜신 것 같습니다.
> 몇가지 코멘트 남겨놓았고 /user/list 접근시 model 셋팅이 안되고있어 재확인 부탁드릴게요~!

@htrucci 어떤 환경으로 테스트하셨나요? 저는 잘 동작합니다. 회원가입을 해야 /user/list에 내용이 나오는데 회원가입후에도 리스트가 출력되지 않나요?

dongjoon1251 commented on src/main/java/controller/UserController.java:
```diff
@@ -31,30 +32,23 @@ private User getUserFromRequest(HttpRequest request) {
-                .getCookies()
-                .stream()
-                .anyMatch(cookie -> cookie.getName().equals(LOGIN_COOKIE_NAME) && cookie.getValue().equals("true"));
+    private boolean isLogined(HttpSession httpSession) {
```
@htrucci  HttpSession이 값은 가지고 있지만 로그인 여부를 확인시켜주는 역할은 아닌 것 같습니다. 사용자가 HttpSession을 가져다가 사용한다고 했을 때, 세션의 어떤 필드를 기준으로 로그인 여부를 확인할지 체크하는 부분은 세션과는 분리가 되어야 한다고 생각합니다. 이전단계 미션에서도 말씀드렸지만, 인증이 필요한 요청들에 대해서 따로 필터링 하는 AuthenticationFilter를 사용자가 정의해서 사용하는게 더 낫다고 생각합니다.

dongjoon1251 commented on src/main/java/webserver/Dispatcher.java:
```diff
@@ -56,14 +67,21 @@ private void dispatchModelAndView(HttpResponse response, Model model, String vie
+        if (requireRequestAndResponseAndModelAndSession(method)) {
+            String viewPath = invokeMethod(request, method, response, model, httpSession);
+            dispatchModelAndView(response, model, viewPath);
+        }
```
HttpSession을 파라미터로 받던 부분을 삭제했습니다. Model 도 안받게 하려고 수정하다보니 return type이 어떤지 또 분기를 타야해서 이부분은 유지시켰습니다.

htrucci reviewed (approved):
많이 부족한 피드백임에도 적극적으로 반영해주셔서 감사합니다. ^^👏 
3단계 모든 요구사항이 잘 반영되어 동작되는 것 확인하였고 머지하도록 하겠습니다.
고생하셨고 다음 미션도 화이팅입니다~! 🎉  🎉

baejunghyun36 committed:
feat : Uri 클래스 구현

baejunghyun36 committed:
feat : HttpVersion 클래스 구현

baejunghyun36 committed:
feat : Method 객체를 별도로 구성
- Uri의 변경사항도 함께 반영

baejunghyun36 committed:
chore : Uri에서 Method를 분리

baejunghyun36 committed:
feat : Uri 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
feat : HttpVersion 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
feat : RequestLine 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
feat : Uri 검증 로직 추가

baejunghyun36 committed:
feat : HttpHeaders 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
test : HttpHeaders 테스트 보완

baejunghyun36 committed:
feat : RequestLine 생성 방식 변경

baejunghyun36 committed:
fix : inputstream 대신 bufferedreader를 주입받는 방식으로 변경 1

baejunghyun36 committed:
fix : inputstream 대신 bufferedreader를 주입받는 방식으로 변경 2

baejunghyun36 committed:
feat : httprequest 클래스 생성

baejunghyun36 committed:
chore : 불필요한 주석 제거

baejunghyun36 committed:
chore : 테스트 클래스 접근지정자 제거

baejunghyun36 committed:
feat : requestbody 관련 로직 추가
- FormUrlEncoded 방식 파싱 로직 추가
- HttpHeaders의 get 메서드 원형 변경
- HttpRequest 구성 요소인 RequestBody 정의
- RequestBodyParsingStrategy의 메서드 분리

baejunghyun36 committed:
refactor : 패키지 구조 수정

baejunghyun36 committed:
feat : StatusLine 클래스 구현

baejunghyun36 committed:
chore : static import로 변경

baejunghyun36 committed:
fix : HttpResponse 구조 변경

baejunghyun36 committed:
docs : 기능 목록 작성

baejunghyun36 committed:
feat : 컨트롤러 등록 구조 작성

baejunghyun36 committed:
feat : 웹 서버 베이스 기능 구현

Commit:
chore : HttpMethod -> Method, 미완성 테스트 코드 완성

Commit:
refactor : WebApplication 구조 변경 및 비효율 구문 변경
- 별도 메서드 분리 및 else 제거
- 불필요한 테스트 제거
- 매 루프마다 UrlDecoding하지 않고 한 번에 Decoding 후 처리로 변경

Commit:
refactor : 매직넘버에 이름을 부여
- 컨트롤러 중복으로 서버 기동 불가 문제와 함께 처리

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -23,32 +26,10 @@ public void run() {
-        } catch (IOException e) {
+            HttpRequest httpRequest = new HttpRequest(in);
+            HttpResponse httpResponse = new HttpResponse(out);
+            RequestMapper.doService(httpRequest, httpResponse);
```
Request Handler 코드 깔끔하네요 😄

sah3122 commented on src/main/java/webserver/common/HttpHeaders.java:
```diff
@@ -0,0 +1,64 @@
+        return Optional.ofNullable(headers.get(key));
+    }
+
+    private final class Parser {
```
Parser 를 별도의 클래스로 분리해준 이유가 있을까요 ?

bigpel66 replied:
HttpHeaders는 헤더 데이터를 갖는 용도로 사용되고, 파싱의 역할을 갖는 것이 적절한지 고민되다보니 별도 클래스로 구성하게 되었습니다.

이 때 각 객체를 만들어주는 범용적인 단일 Parser를 두려고 하기도 했지만, 각각 파싱 방법이 달라서 여러 Parser가 생길 것이라고 생각했었습니다.

각 객체 내부에 이너 클래스로 파서를 갖게 되면 응집력도 높이면서, 별도 역할로 구분할 수 있겠다 싶었습니다.

sah3122 replied:
현재 코드는 Parser를 객체로 생성하지 않고 private 메서드로만 분류하여도 크게 무리는 없을것 같아 남겨둔 코멘트 였습니다. 
별도의 역할을 분리하는것도 좋은 방법이라고 생각해요 😄

sah3122 commented on src/main/java/webserver/common/HttpVersion.java:
```diff
@@ -0,0 +1,104 @@
+        return Objects.hash(raw);
+    }
+
+    private final class Parser {
```
다른 코멘트를 확인해주세요 😄

sah3122 commented on src/main/java/webserver/common/HttpVersion.java:
```diff
@@ -0,0 +1,104 @@
+    private static final int MAJOR_INDEX = 0;
+    private static final int MINOR_INDEX = 1;
+
+    private String scheme;
```
Protocol 을 나타내는 필드 일까요 ? 
Protocol을 나타내는 필드가 맞다면, Version 객체와 분리가 되는것도 좋을것 같아요 😄

bigpel66 replied:
정의한 Scheme은  HTTP/1.1을 예시로 HTTP라는 값이 저장되는 필드입니다.

말씀해주신 방법 대로면 별도의 Protocol을 정의하고, HttpVersion이 Protocol을 갖는 방식으로 이해하면 될까요?

혹시 어떤 이유로 분리되면 좋은지 견해도 궁금합니다!

sah3122 replied:
Request의 프로토콜을 조회하기 위한 기능이 필요할수 있을것 같아 남겨드린 코멘트 입니다 :) 
최대한 HTTP 스펙을 지원하기 위함이라고 봐주시면 될것 같아요 !

bigpel66 replied:
말씀해주신 부분도 아래 코멘트의 QueryString과 함께 반영해보도록 하겠습니다.
감사합니다.

sah3122 commented on src/main/java/webserver/controller/Controller.java:
```diff
@@ -0,0 +1,9 @@
+import webserver.response.HttpResponse;
+
+@FunctionalInterface
+public interface Controller {
```
Controller 클래스 정의 👍

Commit:
chore : 예외 문구 대소문자 변경, header 값 직접 추출 불가 변경

sah3122 commented on src/main/java/webserver/controller/RequestMapper.java:
```diff
@@ -0,0 +1,99 @@
+        staticService(req, res);
+    }
+
+    private static void staticService(HttpRequest req, HttpResponse res) {
```
staticService 를 RequestMapper에서 정의하지 않고 정적리소스를 전달하는 책임을 가지는 클래스를 정의해보는건 어떨까요 ?

bigpel66 replied:
네 이 부분은 다른 나머지 컨트롤러를 정의하면서 함께 리팩토링 진행해보도록 하겠습니다.

sah3122 commented on src/main/java/webserver/controller/RequestMapper.java:
```diff
@@ -0,0 +1,99 @@
+        try {
+            byte[] body = getBody(req);
+            String contentType = getContentType(req);
+            res.response200(body, contentType);
```
HttpResponse 객체를 더 활용해볼수 있을것 같아요. 
ex) Header를 전달하여 필요한 값을 설정

bigpel66 replied:
말씀해주신대로 Request의 Header를 전달받아 Response의 헤더를 설정하는 식의 더 많이 활용하는 쪽으로 구성해보겠습니다.

sah3122 commented on src/main/java/webserver/request/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class FormUrlEncodedParsingStrategy implements RequestBodyParsingStrategy {
```
파싱 전략 클래스 정의 👍

sah3122 commented on src/main/java/webserver/request/ParsingStrategyFactory.java:
```diff
@@ -0,0 +1,12 @@
+package webserver.request;
+
+import org.springframework.http.MediaType;
```
이번 미션에서는 Spring의존성을 제거하여 보세요 😄

bigpel66 replied:
넵 리팩토링 진행하면서 Spring의 의존성도 함께 제거하도록 하겠습니다.

sah3122 commented on src/main/java/webserver/request/ParsingStrategyFactory.java:
```diff
@@ -0,0 +1,12 @@
+    public static RequestBodyParsingStrategy create(String contentType, String contents) {
+        if (MediaType.APPLICATION_FORM_URLENCODED_VALUE.equals(contentType)) {
+            return new FormUrlEncodedParsingStrategy(contents);
+        }
```
ParsingFactory에 content-type에 관한 유효성 검사가 존재합니다. 
하위 구현체들이 content-type의 유효성 검사를 가지지 않게 하려면 구현 전략들의 생성을 막아보는것도 좋을것 같은데 도전해보세요 😄

bigpel66 replied:
죄송하게도 말씀해주신 부분을 이해해보려고 했는데 어떤 말씀이신지 잘 이해가 안 됩니다.

1) 유효성 검사의 의미
Strategy의 다형성을 보장하면서 생성해주는 역할을 Factory에 위임하고, Content-Type에 맞는 객체를 생성하기 위해 타입을 확인하는 부분이 
```
        if (MediaType.APPLICATION_FORM_URLENCODED_VALUE.equals(contentType)) {
            return new FormUrlEncodedParsingStrategy(contents);
        }
```
였습니다.

"적절한 타입으로 생성하기 위해 확인한다"라는 개념도 피드백을 주신 것처럼 유효성 검사라고 표현될 수 있는 것인지 궁금합니다. 

** 유효성에 대한 제 표현 범위가 더 좁은가 싶어서요...
** 아니면 유효성 검사를 진행할 수 있는 부분인데 그게 미약해서 말씀해주신 건가 싶기도 하네요.

2) 의도 파악
주신 피드백이 혹시 하위 구현체가 팩토리를 통해서만 생성되어야 한다는 의미일까요?
유효성 검증을 팩토리에서만 진행하고, 하위 구현체가 유효성 검증을 갖지 않도록하면서 생성을 막는다는 것이 그렇게 이해가 되어서요.

sah3122 replied:
의도 파악에 적어주신 내용이 맞습니다 😄 
팩토리 클래스를 사용한다는것은 생성 로직이 복잡하거나 생성자를 한곳에서 관리하기 위함으로 볼수 있습니다. 
하위 구현체에 유효성 검사 로직이 포함되지 않는다면, 팩토리 클래스를 통하지 않는 객체 생성은 요구사항을 만족하지 못하는 불완전한 객체가 생성될수 있습니다. 
이러한 객체가 생성되지 않기 위해 팩토리 클래스로만 객체 생성이 가능하게 하는 방법과 하위 구현체도 유효성 검사를 추가하는 방법이 있을것 같아요 😄

bigpel66 replied:
여러 방법을 생각해보다가 결국 하위 구현체를 private 접근 지정자로 Factory 내부의 이너 클래스로 정의하면서 생성을 막았는데요.

>하위 구현체에 유효성 검사 로직이 포함되지 않는다면

이라고 말씀해주신걸로 보아, 하위 구현체에서의 유효성 검증으로 팩토리에서의 생성이 아니면 객체가 만들어지지 않도록 강제하는 방법이 있는 거 같은데 혹시 맞을까요?

-> 만일 맞다면 하위 구현체에서는 어떤 식으로 강제할 수 있는지 궁금합니다.
-> 아니라면 어떤 류의 유효성 검증이 작성될 수 있는지 궁금합니다.

sah3122 commented on src/main/java/webserver/request/RequestBody.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Map;
+
+public final class RequestBody {
+    private Map<String, Object> contents;
```
Request body가 항상 Map 형태의 값을 가지고 있지는 않을것 같아요. 😄

bigpel66 replied:
이 부분이 조금 어렵던데 단순히 URL_ENCODED, JSON, XML 등의 데이터를 주고 받을 때만 생각하다보니 Map<String, Object>를 떠올렸던거 같습니다.

혹시 파일과 같은 형태의 데이터를 요청으로 받는 경우엔 위 자료구조를 이용하기 애매할 거 같은데, 말씀해주신 부분이 이런 사례일까요?
만일 맞다면, 어떤 식으로 RequestBody를 구성하는 것이 적절할까요?

sah3122 replied:
저도 깊게 고민해보지는 않은 부분인데, 현재 드는 생각은 RequestBody의 값은 Content-Type에 따라 성격이 정해지는것으로 정의할수 있을것 같아요. 
남겨주신 의견에 제 생각을 더하자면 RequestBody의 인터페이스를 정의하고, Content-Type에 따른 Parser를 따로 분류 하는것도 확장이 가능한 구조가 될것 같네요 !

bigpel66 replied:
RequestBody를 인터페이스를 만들고 각각의 타입을 갖는 구현체를 만드는 방법으로 고려해봤습니다.

RequestBody의 Content를 직접적으로 이용해야하는 로직이 있다면, 공통되는 타입으로 받아내도록 정의해야하는 것으로 이해했는데요.
이 과정에서 공통되는 타입으로 넣으면서, 비즈니스 로직에는 원본 타입에 맞춰서 각 타입에 맞는 로직을 구현해야하는 부분이 문제가 되는 거 같습니다.

이를 해결하기 위해 Reflection이 떠오르기도 했고, 미리 정해둔 각 타입들을 로직에서 활용할 수 있도록 풀어내주는 중간 매개체가 필요할 것 같은데요.
매번 Reflection을 수행하는 것은 그다지 적절해보이지 않고, 그렇다고 각 타입에 맞는 중간 매개체를 구성하려니 너무 큰 작업이라는 생각이 듭니다.

후자가 맞는 거 같은데, Map<String, Object>가 큰 문제가 없다면 미션 진행 중에는 이를 이용해도 될까요?
Parser, Inteferface, Factory 등을 수정해야 되는데, 처음부터 잘 정할걸 싶기도하고 잘 정하는게 중요하구나 싶기도 하네요😢

그래도 한 번 짚어주신 부분을 고민해보고 프레임워크 코드도 본게 도움이 많이 된 거 같습니다.

sah3122 replied:
깊게 고민해주신 부분이 느껴지네요 👍 
말씀 주신것처럼 이번 미션에서는 ` Map<String, Object>`형태로 진행해도 큰 문제가 없다고 생각합니다 😄

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,57 @@
+import java.io.IOException;
+import java.io.OutputStream;
+
+public final class HttpResponse {
```
HttpResponse 가 조금더 상세한 값을 가지고 있어도 될것 같습니다 😄

bigpel66 replied:
StatusLine, HttpHeaders, ResponseBody로 구성해보도록 하겠습니다!

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,57 @@
+        this.dos = new DataOutputStream(out);
+    }
+
+    public void response200(byte[] body, String contentType) {
```
코멘트 남겨주신것처럼 render의 역할을 제거하기 위해 리팩토링 해보세요 😄

bigpel66 replied:
별도 객체로 분리하여 HttpResponse를 받아서 OutputStream에 쓰기 작업을 수행하는 Renderer를 구성하도록 하겠습니다.

Commit:
refactor : 계층 별 캡슐화
- HttpHeaders에서 접근하여 값을 빼내지 않도록 변경

sah3122 reviewed (changes requested):
WAS 미션 구현 잘해주셨습니다 👍 
몇가지 생각거리를 남겨두었는데 확인 부탁드려요 !

Commit:
fix : 테스트 코드 통과

Commit:
fix : 쿼리 파라미터 유효성 검증 방식 변경

Commit:
refactor : Factory를 이용하는 방식으로만 파싱 전략 생성 가능하도록 변경

Commit:
refactor : Spring Framework 의존성 제거

Commit:
refactor : HttpResponse 초기 구성
- HttpVersion의 불필요한 퀄리파이어 삭제
- HttpHeaders의 생성 방식 추가
- 기존 HttpResponse를 Render로 전환
- HttpResponse가 StatusLine과 HttpHeaders를 갖도록 변경
- 임시 조치로 추가 변경 필요 (StatusCode 정의, Renderer 리팩토링 등)

Commit:
refactor : Enum 리패키징

Commit:
feat : StatusCode를 정의

Commit:
feat : ResponseBody를 정의하고 HttpResponse가 이를 갖도록 추가

Commit:
test : 테스트 코드 누락된 부분 커밋

Commit:
feat : HttpResponse가 Render 되도록 로직 구현

Commit:
refactor : RequestMapper가 갖고 있던 StaticServing을 별도 객체로 분리

Commit:
feat : 사용자 회원 가입 로직 구현 (GET / POST)

bigpel66 commented:
* RequestMapper 정적 리소스를 전달하는 별도 Controller를 구성
* HttpResponse의 내부 요소를 Renderer와 분리
* Spring 의존성을 제거

크게는 이렇게 3가지 피드백들을 반영하여, 기능 완성한 것들 다시 리뷰 요청 드리게 되었습니다.
리뷰 꼼꼼히 봐주셔서 정말 감사합니다.

즐거운 휴일 보내셨길 바랍니다.

Commit:
refactor : 페어의 1차 피드백을 반영
- HttpVersion이 매번 생성되지 않도록 static 객체 정의
- HttpVersion의 파싱을 Parser가 담당하도록 변경
- Method의 문자열이 아니라 객체를 활용
- Headers의 생성 방식을 추가 및 테스트

Commit:
docs: 기능 목록 업데이트

Commit:
refactor : Handlebars 템플릿을 지원하도록 변경

sah3122 commented on src/main/java/webserver/common/Uri.java:
```diff
@@ -0,0 +1,109 @@
+    private static final String FIELD_DELIMITER = "=";
+
+    private String path;
+    private Map<String, Object> params;
```
QueryString 도 별도의 클래스로 정의해보는건 어떨까요 ?

bigpel66 replied:
위 피드백의 답글에 단 것처럼 RequestBody에 대한 내용과 추가로 QueryString까지 별도로 구성해보도록 하겠습니다.

sah3122 commented on src/main/java/webserver/controller/GetUserJoinController.java:
```diff
@@ -0,0 +1,23 @@
+import webserver.request.HttpRequest;
+import webserver.response.HttpResponse;
+
+public class GetUserJoinController implements Controller {
```
비즈니스 로직 정의 👍

sah3122 commented on src/main/java/webserver/controller/RequestMapper.java:
```diff
@@ -0,0 +1,60 @@
+            map.get(entry).doService(req, res);
+            return;
+        }
+        map.get(new Entry(Method.GET, DEFAULT_FILE_SERVE_PATH)).doService(req, res);
```
정의되지 않은 request Path의 경우 NotFound 응답을 줄수 있을까요 ?😄

bigpel66 replied:
피드백 주신 내용을 의식하여 NotFound를 보내고 있습니다!

repsonse를 주는 부분이 RequestMapper의 역할은 아니라고 생각해서 여기서 기재하진 않았고, 

```
map.get(new Entry(Method.GET, DEFAULT_FILE_SERVE_PATH)).doService(req, res);
```
에서 사용되는 StaticServingController의 마지막 구문에서 404를 내도록 로직을 구성하긴 했었습니다.

<img width="1840" alt="스크린샷 2024-04-11 오후 5 52 11" src="https://github.com/next-step/jwp-was-kakao/assets/47660338/32c84f2d-a5c9-4bbb-88d2-e92d73e3617e">

bigpel66 replied:
이와 관련하여 Template을 Render하는 코드에서 확장자를 제외한 파일 이름만으로 매칭하다보니, 확장자명이 다르거나 존재하지 않아도 Not Found가 발생하지 않는 문제가 있었습니다.

수정하여 함께 반영하겠습니다.

sah3122 commented on src/main/java/webserver/controller/StaticServingController.java:
```diff
@@ -0,0 +1,66 @@
+        try {
+            return getTemplatesBody(httpRequest);
+        } catch (FileNotFoundException | NullPointerException e) {
+            return getStaticBody(httpRequest);
```
try-catch로 요청에 해당하는 리소스를 탐색하기 보단, static resource의 uri path 패턴을 관리하여 보는건 어떨까요 ?

sah3122 commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,97 @@
+    private final RequestLine requestLine;
+    private final HttpHeaders httpHeaders;
+    private final RequestBody requestBody;
+    private Object attribute;
```
해당 필드는 어떤 필드 일까요 ? 
값을 바인딩하는 코드가 없어 질문드려요 😄

bigpel66 replied:
아 이 부분이 2단계에 해당하는 부분인걸 놓치고 step1에 반영하게 되었습니다.
강의 듣다보니 이 부분을 놓쳤나 해서 만들고 보니 뒤늦게 step2라는걸 알게 되었습니다.
템플릿 엔진을 이용하는 부분에서 구체화 하여 활용해보도록 하겠습니다.

sah3122 commented on src/main/java/webserver/request/RequestBodyParsingStrategyFactory.java:
```diff
@@ -0,0 +1,47 @@
+
+import static webserver.enums.MediaType.APPLICATION_FORM_URLENCODED;
+
+public final class RequestBodyParsingStrategyFactory {
```
RequestBodyParser를 구현해주셨네요 👍

sah3122 reviewed (changes requested):
피드백 반영 잘해주셨네요 👍 
몇가지 사소한 코멘트를 추가해두었는데 다음 미션을 진행하기전 확인해보면 좋을것 같아요. 🙏 
궁금한부분이 있으면 편하게 DM 주세요 !

Commit:
fix : template 엔진의 확장자에 대해서도 매칭 여부를 검증

Commit:
refactor : Path 객체를 보유하도록 변경
- Uri가 갖고 있는 Path를 String 타입이 아닌 별도 객체로 정의
- Templates와 Static의 파일 관련 동작을 FileIOUtils의 역할로 분리

Commit:
refactor : QueryString 객체를 별도로 정의

bigpel66 commented:
말씀해주신 부분들 중 크게 

- Path 객체를 생성
- QueryString 객체를 생성
- Try-Catch가 아닌 파일 경로를 통해 존재 여부 확인

을 반영했습니다.

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
사소한 피드백들도 깊게 고민해주셔서 저도 도움이 많이 되었습니다. 🙏 
다음 미션도 기대할게요 😄
